services/redeployment/service_configmap.go

```diff
@@ -0,0 +1,123 @@
+package redeployment
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	ctrclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type configMapService struct {
+	client    ctrclient.Client
+	configMap *corev1.ConfigMap
+	ctx       context.Context
+
+	pods         []corev1.Pod
+	matchingPods map[string]corev1.Pod
```

Comment:
this model sketches me out a bit, because I'm not sure if it's robust to strange concurrency patterns. 

I know controllers are typically single threaded, but it still feels safer to go with a more functional rather than stateful approach.

Reply:
controller by design must react to resource changes and reflect a new object state. Allows you to control how stateful or complex deployments are executed. Generally, it does like this. Watches secret or config map changes and deletes affected Pods (changes Pod state). The controller manager uses cached Pods so the solution is fast and robust.

Reply:
Even if the controller would run multiple threads the instance of this object would not be shared across reconcile loop executions since service is created on every reconcile. This is only a minimal optimization, not to list pods multiple times. https://github.com/pluralsh/plural-operator/pull/8/files#diff-1e03cab057dd3214946fba681d6c19d581f72f23f6e6c15a90f448aa43947d44R48-R63

Reply:
Ok sounds good then